File: MVC_O_DiceGame/DiceGame_ruby/model/DiceGame.rb
require './model/Dice.rb'

module Model
  class DiceGame

    def initialize
      @d1 = Model::Dice.new
      @d2 = Model::Dice.new
      @d3 = Model::Dice.new
      @observers = []
    end

    def add_subscriber subscriber
      @observers << subscriber
    end

    def roll dice
      dice.roll
      @observers.each do |o|
        o.rolled_dice dice.face_value
      end
    end

    def play
      roll @d1
      sleep 2
      roll @d2
      sleep 2
      roll @d3

      @d1.face_value + @d2.face_value + @d3.face_value == 9
    end

  end
end

File: MVC_O_DiceGame/DiceGame_java/model/DiceGame.java
package model;

public class DiceGame {
  private Dice m_d1;
  private Dice m_d2;
  private Dice m_d3;
  
  private java.util.ArrayList<IDiceRolledObserver> m_subscribers;
  
  public DiceGame() {
    m_d1 = new Dice();
    m_d2 = new Dice();
    m_d3 = new Dice();
    m_subscribers = new java.util.ArrayList<IDiceRolledObserver>();
  }
  
  public void addSubscriber(IDiceRolledObserver a_sub) {
    m_subscribers.add(a_sub);
  }
  
  private void rollDice(Dice a_dice) {
    a_dice.roll();
    for(IDiceRolledObserver obs : m_subscribers) {
      obs.DiceRolled(a_dice.getValue());
    }
    try {
      Thread.sleep(2000);
    } catch (Exception e) {
    }
  }
  
  public boolean play() {
    
    rollDice(m_d1);
    rollDice(m_d2);
    rollDice(m_d3);
    
    return m_d1.getValue() + m_d2.getValue() + m_d3.getValue() == 9;
  }
}